docs/madr/decisions/055-default-hostname-generators.md

```diff
@@ -0,0 +1,218 @@
+# Default HostnameGenerators
+
+* Status: accepted
+
+## Context and Problem Statement
+
+We introduced HostnameGenerator for MeshService and MeshExternalService which is also compatible with upcoming MeshMultizoneService.
+This MADR proposes default generators taking into account different scenarios.
+
+## Considered Options
+
+Template and matching:
+* Proposed options
+
+Shipping:
+* Document
+* Create
+
+Suffix:
+* `.svc.{{ type }}.{{ zone }}`
+* `.{{ zone }}.{{ type }}`
+
+## Decision Outcome
+
+Template and matching - Proposed options.
+Shipping - short term document. Once we get all prerequisites and MeshService/MeshExternalService/MeshMultizoneService goes GA, we create by default.
+Suffix - ?
+
+## Pros and Cons of the Options - Template and matching
+
+### Prerequisites
+
+**Labels**
+* `kuma.io/origin`. The value is either `zone` or `global`.
+  The initial idea was to preserve the label between Zones, so that `kuma.io/origin: zone` in Zone A means that it's also the same on Global and on Zone B (if resource is synced).
+  However, we diverged from this, so that Zone A has `kuma.io/origin: zone`, Global has `kuma.io/origin: zone` and Zone B has `kuma.io/origin: global` (because it was synced from global).
+  We could adjust the implementation to initial assumptions, but it requires careful handling. If we were to do this we would have to introduce something like `kuma.io/synced: "true"`.
+  We found existing state to be useful for hostname generators, and we did not find good use case to leverage this label preservation across zones.
+* `kuma.io/env`. The value is either `kubernetes` or `universal`.
+  Introducing this label is out of scope of this MADR.
+
+**Display name**
+`{{ .DisplayName }}` resolves to `kuma.io/display-name` if it exists and it fallbacks to name.
+We can remove the fallback if we add `kuma.io/display-name` to Universal resources by default.
+It's better than `{{ label "kuma.io/display-name" }}`, because looking at Kube resources it's annotation, not a label, so it might be confusing for users.
+
+### MeshService
+
+#### Local Kubernetes MeshService
+
+It's rather unlikely for user to use custom hostnames to consume MeshService that is local in the Kubernetes cluster.
+Therefore, no default HostnameGenerator is needed for it.
+
+#### Synced Kubernetes MeshService
+
```

Comment:
Wondering whether we need a separate generator for headless services here. Since the name for headless is sort of useless

Reply:
We do, let me update MADR

Reply:
Random question, how do we manage names of MES for headless svcs? How do we ensure adding the suffix with the pod offset doesn't lead to too long names?

Reply:
> MES for headless svcs?

MeshExternalService for Headless svcs?

> How do we ensure adding the suffix with the pod offset doesn't lead to too long names

Do you mean to not go over 63 char of DNS label (=subdomain)?

Reply:
https://github.com/kumahq/kuma/blob/16c7cb0c137ebc084d960b81e4d8bce21588a359/pkg/plugins/runtime/k8s/controllers/meshservice_controller.go#L206

Reply:
So we truncate ok. Is this something k8s does as well? I'm just scared of creating new issues (This is actually parallel to this MADR and it might be better to track as its own issue)

Reply:
Yeah: https://github.com/kubernetes/kubernetes/blob/29e4f5a893bf47c608aa8593ae6e67d20b0fd775/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L49